tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_runner.cc

```diff
@@ -0,0 +1,306 @@
+      params->common.epsilon,                        //
+      se::dnn::ActivationMode::kNone,                //
+      &output_data,                                  //
+      /*batch_mean=*/&null_device_ptr,               //
+      /*batch_var=*/&null_device_ptr,                //
+      /*saved_mean=*/&params->output_mean,           //
+      /*saved_inv_var=*/&params->output_inv_stddev,  //
+      /*is_training=*/true,                          //
+      /*var_to_inv_var=*/nullptr,                    //
+      /*inv_var_to_var=*/nullptr,                    //
+      /*reserve_space_allocator=*/nullptr,           //
+      /*workspace_allocator=*/nullptr);
+}
+
+template <typename ElemType>
+void RunCudnnBatchNormBackwardImpl(CudnnBatchNormBackwardParams* params,
+                                   se::Stream* stream) {
+  se::DeviceMemory<float> null_device_ptr(nullptr);
+  auto output_grad_data = se::DeviceMemory<ElemType>(params->output_grad_data);
+  stream->ThenBatchNormalizationBackward(
+      se::DeviceMemory<ElemType>(params->grad_output),     //
+      se::DeviceMemory<ElemType>(params->common.operand),  //
+      params->common.scale,                                //
+      params->mean,                                        //
+      params->inv_stddev,                                  //
+      params->common.operand_desc,                         //
+      params->common.scale_offset_desc,                    //
+      params->common.epsilon,                              //
+      &output_grad_data,                                   //
+      &params->output_grad_scale,                          //
+      &params->output_grad_offset,                         //
+      /*reserve_space_allocator=*/nullptr,                 //
+      /*workspace_allocator=*/nullptr);
+}
+
+}  // namespace
+
+Status RunCudnnBatchNormForwardInference(
+    const HloInstruction* batchnorm, se::DeviceMemoryBase operand,
+    se::DeviceMemoryBase output, se::DeviceMemory<float> scale,
+    se::DeviceMemory<float> offset, se::DeviceMemory<float> mean,
+    se::DeviceMemory<float> variance, float epsilon, int64 feature_index,
+    se::Stream* stream) {
+  CudnnBatchNormForwardInferenceParams inference_params;
+  AssignCommonParams(batchnorm, &inference_params.common, operand, scale,
+                     epsilon, feature_index);
+  inference_params.offset = offset;
+  inference_params.mean = mean;
+  inference_params.variance = variance;
+  inference_params.output = output;
+
+  PrimitiveType output_primitive_type = batchnorm->shape().element_type();
+  switch (output_primitive_type) {
+    case F16:
+      RunCudnnBatchNormForwardInferenceImpl<Eigen::half>(&inference_params,
+                                                         stream);
+      break;
+    case F32:
+      RunCudnnBatchNormForwardInferenceImpl<float>(&inference_params, stream);
+      break;
+    default:
+      Unimplemented("Primitive type not implemented for \"%s\" ",
```

Comment:
Should this be `return Unimplemented(...)`?

Reply:
Same below.

Reply:
Damn! I think I'm not on my game today :(